### PR TITLE
[CAZ-2846] Leeds Taxi Weekly Selection - Javascript enabled

### PR DESCRIPTION
**Pull request by pkalamarski**
## Motivation and Context

https://eaflood.atlassian.net/browse/CAZ-2846

## Description

Added Javascript pack that contains a date picker integrated with inputs, validates and displays available pay dates. Also highlights the week that the user could pay for.

## How Has This Been Tested?

Manually on Chrome 83 and IE11, rspecs

## Screenshots (if appropriate):
image

image

Entering correct value/mouse hover over an available day:
image

Clicking/entering invalid date:
image

End day notice:
image

Date out of range:
image

Date already paid:
image

All paid:
image

## Checklist:

- [x] It contains only changes required by issue (does not contain other PR)
- [x] Includes the link to an issue (if apply)
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: app/controllers/dates_controller.rb [hunk 1]
```diff
@@ -354,4 +358,40 @@ def add_confirm_charge_today_to_session
...
+  end
+
+  def date_in_range?(date)
+    return unless date
+
+    start_day = Time.zone.now.beginning_of_day - 6.days
+    end_day = Time.zone.now.beginning_of_day + 6.days
+
+    available_dates_range = start_day...end_day
+    available_dates_range.cover?(Time.zone.parse(date))
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:379]: https://gist.github.com/julian7/4055746
better to use `between?`

File: spec/services/dates/validate_selected_weekly_date_spec.rb [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+require 'rails_helper'
+
+RSpec.describe Dates::ValidateSelectedWeeklyDate do
+  subject(:service) { described_class.new(params: date_partials) }
+
+  context 'when date is in correct format and is in range' do
+    year = Time.current.year
+    month = Time.current.month
+    day = Time.current.day
```
Comment (likeuwill) [inline spec/services/dates/validate_selected_weekly_date_spec.rb:11]: move it to let variables and use it in each context

File: app/controllers/dates_controller.rb [hunk 3]
```diff
@@ -227,13 +229,14 @@ def select_weekly_date # rubocop:disable Metrics/AbcSize
   # * +charge+ - lack of VRN redirects to {enter_details}[rdoc-ref:VehiclesController.enter_details]
   #
   def confirm_date_weekly
-    dates = params[:dates]
-    if dates && check_already_paid_weekly(dates)
+    service = Dates::ValidateSelectedWeeklyDate.new(params: params)
```
Comment (tkaczu) [inline app/controllers/dates_controller.rb:232]: Nice one!

File: spec/services/dates/validate_selected_weekly_date_spec.rb (hunk 2 above)
Comment (tkaczu) [reply spec/services/dates/validate_selected_weekly_date_spec.rb:11]: Like Sergi advised, change to let statements, and only overide them in each context and do not overide `date_partials` then.
```suggestion
    let(:year) { Time.current.year }
    let(:month) { Time.current.month }
    let(:day) { Time.current.day }
```

File: spec/services/dates/validate_selected_weekly_date_spec.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      expect(service.error).to eq(I18n.t('paid', scope: 'dates.weekly'))
+    end
+  end
+
+  context 'when date is in correct format but is out of range' do
+    year = '1842'
+    month = '5'
+    day = '14'
+
+    let(:date_partials) do
```
Comment (tkaczu) [inline spec/services/dates/validate_selected_weekly_date_spec.rb:43]: If You follow our comments about assigning year, month and day in lets statements, You could ommit `let(:date_partials)` in each context.

File: app/controllers/dates_controller.rb [hunk 5]
```diff
@@ -227,13 +229,14 @@ def select_weekly_date # rubocop:disable Metrics/AbcSize
...
+
+    dates = [service.start_date]
+    if service.valid? && check_already_paid_weekly(dates)
       SessionManipulation::CalculateTotalCharge.call(session: session, dates: dates, weekly: true)
```
Comment (likeuwill) [inline app/controllers/dates_controller.rb:236]:
```
if service.valid? && check_already_paid_weekly([service.start_date])
     SessionManipulation::CalculateTotalCharge.call(session: session, dates: [service.start_date], weekly: true)
```

File: spec/requests/dates_controller/confirm_date_weekly_spec.rb [hunk 6]
```diff
@@ -17,7 +17,17 @@
   end
   let(:vrn) { 'CU123AB' }
   let(:la_id) { SecureRandom.uuid }
-  let(:dates) { { 'dates' => ['2019-11-01'] } }
+  let(:date_partials) { { 'date_year' => '2019', 'date_month' => '11', 'date_day' => '01' } }
```
Comment (likeuwill) [inline spec/requests/dates_controller/confirm_date_weekly_spec.rb:20]: just `params`

File: spec/services/dates/validate_selected_weekly_date_spec.rb [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+RSpec.describe Dates::ValidateSelectedWeeklyDate do
+  subject(:service) { described_class.new(params: date_partials) }
+
+  context 'when date is in correct format and is in range' do
+    year = Time.current.year
+    month = Time.current.month
+    day = Time.current.day
+
+    let(:date_partials) do
```
Comment (likeuwill) [inline spec/services/dates/validate_selected_weekly_date_spec.rb:13]: same, just `params`

**Pull request merged**